api/services/content/content.proto

```diff
@@ -0,0 +1,156 @@
+
+	// Offset specifies the number of bytes in the blob to begin the read. If
+	// zero or less, the read will be from the start.
+	int64 offset = 2;
+
+	// size is the total size of the read. If zero, the entire blob will be
+	// returned by the service.
+	int64 size = 3;
+}
+
+// ReadResponse carries byte data for a read request.
+message ReadResponse {
+	int64 offset = 1; // offset of the returned data
+	bytes data = 2; // actual data
+}
+
+
+// WriteAction sets the action for a WriteRequst.
+enum WriteAction {
+	option (gogoproto.goproto_enum_prefix) = false;
+	option (gogoproto.enum_customname) = "WriteAction";
+	
+	// WriteActionWrite sets the action for the write request to write data.
+	//
+	// Any data included will be written at the provided offset. The
+	// transaction will be left open for further writes.
+	//
+	// This is the default.
+	WRITE = 0 [(gogoproto.enumvalue_customname)="WriteActionWrite"];
+
+	// WriteActionCommit will write any outstanding data in the message and
+	// commit the write, storing it under the digest.
+	//
+	// This can be used in a single message to send the data, verify it and
+	// commit it.
+	COMMIT = 1 [(gogoproto.enumvalue_customname)="WriteActionCommit"];
+
+	// WriteActionAbort will release any resources associated with the write
+	// and free up the ref for a completely new set of writes.
+	ABORT = -1 [(gogoproto.enumvalue_customname)="WriteActionAbort"];
+}
+
+// WriteRequest writes data to the request ref at offset. 
+message WriteRequest {
+	// Action sets the behavior of the write.
+	//
+	// When this is a write and the ref is not yet allocated, the ref will be
+	// allocated and the data will be written at offset.
+	//
+	// If the action is write and the ref is allocated, it will accept data to
+	// an offset that has not yet been written.
+	//
+	// If the action is write and there is no data, the current write status
+	// will be returned. This works differently from status because the stream
+	// holds a lock.
+	WriteAction action = 1;
+	string ref = 2;
+	int64 expected_size = 3;
+	string expected_digest = 4 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+	int64 offset = 5;
+	bytes data = 6;
+}
+
+message WriteResponse {
+	// Action contains the action for the final message of the stream. A writer
+	// should confirm that they match the intended result.
+	WriteAction action = 1;
+
+	// Size provides the current "committed" for the Write.
+	string size = 2;
+
+	// Digest, if present, includes the digest up to the currently committed
+	// bytes. If action is commit, this field will be set. It is implementation
+	// defined if this is set for other actions, except abort. On abort, this
+	// will be empty.
+	string digest = 3 [(gogoproto.customtype) = "github.com/opencontainers/go-digest.Digest", (gogoproto.nullable) = false];
+}
+
+message StatusRequest {
+	repeated string refs = 1;
+	repeated string prefix = 2;
```

Comment:
It is a matcher for the status prefix of refs. I haven't fully defined the semantics, but imagine you have a bunch of ref keys like 'foo-sha256:xxxxx`. You could specify the prefix for `foo-` and get the status of all those that match.

I am going to attempt an implementation here in a bit and figure out what actually makes sense for  displaying partial statuses.